File: atlas/media_requests/discovery.py
"""Read-only media-discovery contracts for Project Atlas."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any

from .models import MediaRequestType


class MediaDiscoveryError(ValueError):
    """Raised when a normalized media-discovery contract is invalid."""


class MediaDiscoveryAvailability(str, Enum):
    """Normalized Jellyseerr discovery availability state."""

    NOT_TRACKED = "not_tracked"
    UNKNOWN = "unknown"
    PENDING = "pending"
    PROCESSING = "processing"
    PARTIALLY_AVAILABLE = "partially_available"
    AVAILABLE = "available"
    BLOCKLISTED = "blocklisted"
    DELETED = "deleted"


@dataclass(frozen=True, slots=True)
class MediaDiscoveryItem:
    """One normalized movie or TV discovery result."""

    provider_media_id: str
    media_type: MediaRequestType
    title: str
    availability: MediaDiscoveryAvailability
    year: int | None = None
    overview: str | None = None
    poster_path: str | None = None

    def __post_init__(self) -> None:
        provider_media_id = _numeric_identity(
            self.provider_media_id,
            "provider_media_id",
        )

        media_type = _media_type(
            self.media_type,
            "media_type",
        )

        title = _required_text(
            self.title,
            "title",
        )

        availability = _availability(
            self.availability,
            "availability",
        )

        year = _optional_year(
            self.year,
            "year",
        )

        overview = _optional_text(
            self.overview,
            "overview",
        )

        poster_path = _optional_path(
            self.poster_path,
            "poster_path",
        )

        object.__setattr__(
            self,
            "provider_media_id",
            provider_media_id,
        )
        object.__setattr__(
            self,
            "media_type",
            media_type,
        )
        object.__setattr__(
            self,
            "title",
            title,
        )
        object.__setattr__(
            self,
            "availability",
            availability,
        )
        object.__setattr__(
            self,
            "year",
            year,
        )
        object.__setattr__(
            self,
            "overview",
            overview,
        )
        object.__setattr__(
            self,
            "poster_path",
            poster_path,
        )

    @property
    def request_eligible(
        self,
    ) -> bool:
        """Return whether B3.3 may present a Request action."""

        return (
            self.availability
            is MediaDiscoveryAvailability.NOT_TRACKED
        )

    def to_dict(
        self,
    ) -> dict[str, Any]:
        """Return a deterministic serialized discovery item."""

        return {
            "provider_media_id":
                self.provider_media_id,
            "media_type":
                self.media_type.value,
            "title":
                self.title,
            "year":
                self.year,
            "overview":
                self.overview,
            "poster_path":
                self.poster_path,
            "availability":
                self.availability.value,
            "request_eligible":
                self.request_eligible,
        }


@dataclass(frozen=True, slots=True)
class MediaDiscoveryPage:
    """One normalized Jellyseerr discovery/search page."""

    items: tuple[
        MediaDiscoveryItem,
        ...,
    ]
    page: int
    total_pages: int

    def __post_init__(self) -> None:
        try:
            items = tuple(
                self.items
            )
        except TypeError as exc:
            raise MediaDiscoveryError(
                "items must be iterable"
            ) from exc

        for index, item in enumerate(
            items
        ):
            if not isinstance(
                item,
                MediaDiscoveryItem,
            ):
                raise MediaDiscoveryError(
                    f"items[{index}] must be a "
                    "MediaDiscoveryItem"
                )

        page = _positive_integer(
            self.page,
            "page",
        )

        total_pages = _non_negative_integer(
            self.total_pages,
            "total_pages",
        )

        if items and total_pages == 0:
            raise MediaDiscoveryError(
                "total_pages cannot be zero "
                "when items are present"
            )

        identities = {
            (
                item.media_type,
                item.provider_media_id,
            )
            for item in items
        }

        if len(identities) != len(items):
            raise MediaDiscoveryError(
                "discovery item identities "
                "must be unique within a page"
            )

        object.__setattr__(
            self,
            "items",
            items,
        )
        object.__setattr__(
            self,
            "page",
            page,
        )
        object.__setattr__(
            self,
            "total_pages",
            total_pages,
        )

    @property
    def next_page(
        self,
    ) -> int | None:
        """Return the next provider page when one exists."""

        if (
            self.total_pages == 0
            or self.page >= self.total_pages
        ):
            return None

        return self.page + 1

    def to_dict(
        self,
    ) -> dict[str, Any]:
        """Return a deterministic serialized discovery page."""

        return {
            "items": [
                item.to_dict()
                for item in self.items
            ],
            "page":
                self.page,
            "total_pages":
                self.total_pages,
            "next_page":
                self.next_page,
        }


def _required_text(
    value: object,
    field_name: str,
) -> str:
    if not isinstance(
        value,
        str,
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be text"
        )

    normalized = value.strip()

    if not normalized:
        raise MediaDiscoveryError(
            f"{field_name} is required"
        )

    return normalized


def _optional_text(
    value: object,
    field_name: str,
) -> str | None:
    if value is None:
        return None

    if not isinstance(
        value,
        str,
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be text or null"
        )

    normalized = value.strip()

    return (
        normalized
        if normalized
        else None
    )


def _optional_path(
    value: object,
    field_name: str,
) -> str | None:
    normalized = _optional_text(
        value,
        field_name,
    )

    if normalized is None:
        return None

    if not normalized.startswith(
        "/"
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be a relative provider path"
        )

    return normalized


def _numeric_identity(
    value: object,
    field_name: str,
) -> str:
    if (
        isinstance(value, bool)
        or not isinstance(
            value,
            (
                str,
                int,
            ),
        )
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be text or an integer"
        )

    normalized = str(
        value
    ).strip()

    if not normalized:
        raise MediaDiscoveryError(
            f"{field_name} is required"
        )

    if (
        not normalized.isdigit()
        or int(normalized) <= 0
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be a positive "
            "numeric TMDB identifier"
        )

    return normalized


def _media_type(
    value: object,
    field_name: str,
) -> MediaRequestType:
    if isinstance(
        value,
        MediaRequestType,
    ):
        normalized = value
    elif isinstance(
        value,
        str,
    ):
        candidate = (
            value
            .strip()
            .lower()
            .replace("-", "_")
            .replace(" ", "_")
        )

        try:
            normalized = (
                MediaRequestType(
                    candidate
                )
            )
        except ValueError as exc:
            raise MediaDiscoveryError(
                f"{field_name} is unsupported"
            ) from exc
    else:
        raise MediaDiscoveryError(
            f"{field_name} must be a "
            "MediaRequestType or text"
        )

    if normalized not in {
        MediaRequestType.MOVIE,
        MediaRequestType.TV,
    }:
        raise MediaDiscoveryError(
            f"{field_name} must be movie or tv"
        )

    return normalized


def _availability(
    value: object,
    field_name: str,
) -> MediaDiscoveryAvailability:
    if isinstance(
        value,
        MediaDiscoveryAvailability,
    ):
        return value

    if not isinstance(
        value,
        str,
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be "
            "MediaDiscoveryAvailability or text"
        )

    normalized = (
        value
        .strip()
        .lower()
        .replace("-", "_")
        .replace(" ", "_")
    )

    try:
        return MediaDiscoveryAvailability(
            normalized
        )
    except ValueError as exc:
        raise MediaDiscoveryError(
            f"{field_name} is unsupported"
        ) from exc


def _optional_year(
    value: object,
    field_name: str,
) -> int | None:
    if value is None:
        return None

    if (
        isinstance(value, bool)
        or not isinstance(
            value,
            int,
        )
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be an integer or null"
        )

    current_year = (
        datetime.now(
            timezone.utc
        ).year
    )

    if not (
        1888
        <= value
        <= current_year + 10
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be between "
            f"1888 and {current_year + 10}"
        )

    return value


def _positive_integer(
    value: object,
    field_name: str,
) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(
            value,
            int,
        )
        or value <= 0
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be a positive integer"
        )

    return value


def _non_negative_integer(
    value: object,
    field_name: str,
) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(
            value,
            int,
        )
        or value < 0
    ):
        raise MediaDiscoveryError(
            f"{field_name} must be a "
            "non-negative integer"
        )

    return value

File: apps/portal/features/media/types/discovery.ts
export const MEDIA_DISCOVERY_MEDIA_TYPES = ["movie", "tv"] as const;

export type MediaDiscoveryMediaType = (typeof MEDIA_DISCOVERY_MEDIA_TYPES)[number];

export const MEDIA_DISCOVERY_AVAILABILITIES = [
  "not_tracked",
  "unknown",
  "pending",
  "processing",
  "partially_available",
  "available",
  "blocklisted",
  "deleted"
] as const;

export type MediaDiscoveryAvailability = (typeof MEDIA_DISCOVERY_AVAILABILITIES)[number];

export type MediaDiscoveryItem = Readonly<{
  providerMediaId: string;
  mediaType: MediaDiscoveryMediaType;
  title: string;
  year?: number;
  overview?: string;
  posterPath?: string;
  availability: MediaDiscoveryAvailability;
  requestEligible: boolean;
}>;

export type MediaDiscoveryPage = Readonly<{
  items: readonly MediaDiscoveryItem[];
  page: number;
  totalPages: number;
  nextPage: number | null;
}>;

export type CreateMediaDiscoveryPageInput = Readonly<{
  items: readonly MediaDiscoveryItem[];
  page: number;
  totalPages: number;
}>;

function normalizeRequiredText(value: string, fieldName: string): string {
  const normalized = value.trim();

  if (!normalized) {
    throw new Error(`${fieldName} is required.`);
  }

  return normalized;
}

function normalizeOptionalText(value: string | undefined): string | undefined {
  const normalized = value?.trim();

  return normalized ? normalized : undefined;
}

export function normalizeMediaDiscoveryProviderId(value: string): string {
  const normalized = normalizeRequiredText(value, "providerMediaId");

  if (!/^[0-9]+$/.test(normalized) || Number(normalized) <= 0) {
    throw new Error("providerMediaId must be a positive numeric TMDB identifier.");
  }

  return normalized;
}

export function normalizeMediaDiscoveryType(value: string): MediaDiscoveryMediaType {
  const normalized = value.trim().toLowerCase();

  if (!MEDIA_DISCOVERY_MEDIA_TYPES.includes(normalized as MediaDiscoveryMediaType)) {
    throw new Error("mediaType must be movie or tv.");
  }

  return normalized as MediaDiscoveryMediaType;
}

export function normalizeMediaDiscoveryAvailability(value: string): MediaDiscoveryAvailability {
  const normalized = value.trim().toLowerCase().replace(/[- ]/g, "_");

  if (!MEDIA_DISCOVERY_AVAILABILITIES.includes(normalized as MediaDiscoveryAvailability)) {
    throw new Error("availability is unsupported.");
  }

  return normalized as MediaDiscoveryAvailability;
}

export function normalizeMediaDiscoveryQuery(value: string): string {
  const normalized = normalizeRequiredText(value, "query");

  if (normalized.length > 200) {
    throw new Error("query must contain at most 200 characters.");
  }

  return normalized;
}

export function normalizeMediaDiscoveryPageNumber(value: number): number {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error("page must be a positive integer.");
  }

  return value;
}

function normalizeYear(value: number | undefined): number | undefined {
  if (value === undefined) {
    return undefined;
  }

  const maximumYear = new Date().getUTCFullYear() + 10;

  if (!Number.isInteger(value) || value < 1888 || value > maximumYear) {
    throw new Error(`year must be between 1888 and ${maximumYear}.`);
  }

  return value;
}

function normalizePosterPath(value: string | undefined): string | undefined {
  const normalized = normalizeOptionalText(value);

  if (normalized === undefined) {
    return undefined;
  }

  if (!normalized.startsWith("/")) {
    throw new Error("posterPath must be a relative provider path.");
  }

  return normalized;
}

export function createMediaDiscoveryItem(item: MediaDiscoveryItem): MediaDiscoveryItem {
  const providerMediaId = normalizeMediaDiscoveryProviderId(item.providerMediaId);
  const mediaType = normalizeMediaDiscoveryType(item.mediaType);
  const title = normalizeRequiredText(item.title, "title");
  const year = normalizeYear(item.year);
  const overview = normalizeOptionalText(item.overview);
  const posterPath = normalizePosterPath(item.posterPath);
  const availability = normalizeMediaDiscoveryAvailability(item.availability);

  if (typeof item.requestEligible !== "boolean") {
    throw new Error("requestEligible must be boolean.");
  }

  const expectedEligibility = availability === "not_tracked";

  if (item.requestEligible !== expectedEligibility) {
    throw new Error("requestEligible does not match the discovery availability state.");
  }

  return {
    providerMediaId,
    mediaType,
    title,
    ...(year === undefined ? {} : { year }),
    ...(overview === undefined ? {} : { overview }),
    ...(posterPath === undefined ? {} : { posterPath }),
    availability,
    requestEligible: expectedEligibility
  };
}

export function createMediaDiscoveryPage(input: CreateMediaDiscoveryPageInput): MediaDiscoveryPage {
  const items = input.items.map(createMediaDiscoveryItem);
  const page = normalizeMediaDiscoveryPageNumber(input.page);

  if (!Number.isInteger(input.totalPages) || input.totalPages < 0) {
    throw new Error("totalPages must be a nonnegative integer.");
  }

  const totalPages = input.totalPages;

  if (items.length > 0 && totalPages === 0) {
    throw new Error("totalPages cannot be zero when discovery items are present.");
  }

  if (totalPages > 0 && page > totalPages) {
    throw new Error("page cannot exceed totalPages.");
  }

  const identities = new Set(items.map((item) => `${item.mediaType}:${item.providerMediaId}`));

  if (identities.size !== items.length) {
    throw new Error("Media discovery identities must be unique within a page.");
  }

  const nextPage = totalPages > 0 && page < totalPages ? page + 1 : null;

  return {
    items,
    page,
    totalPages,
    nextPage
  };
}

export function mediaDiscoveryAvailabilityLabel(availability: MediaDiscoveryAvailability): string {
  switch (availability) {
    case "not_tracked":
      return "Not tracked";
    case "unknown":
      return "Unknown";
    case "pending":
      return "Pending";
    case "processing":
      return "Processing";
    case "partially_available":
      return "Partially available";
    case "available":
      return "Available";
    case "blocklisted":
      return "Blocklisted";
    case "deleted":
      return "Deleted";
  }
}
